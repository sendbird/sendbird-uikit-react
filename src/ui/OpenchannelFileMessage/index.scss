@import '../../styles/variables';

.sendbird-openchannel-file-message {
  position: relative;
  display: flex;
  flex-direction: row;
  &:hover {
    @include themed() {
      background-color: t(bg--1-4);
    }
    .sendbird-openchannel-file-message__context-menu {
      display: block;
    }
  }

  .sendbird-openchannel-file-message__left {
    position: relative;
    display: flex;
    width: 52px;
    min-width: 52px;

    .sendbird-openchannel-file-message__left__avatar {
      position: absolute;
      display: block;
      margin-top: 8px;
      margin-left: 24px;
    }
  }

  .sendbird-openchannel-file-message__right {
    position: relative;
    display: flex;
    flex-direction: column;
    margin-left: 12px;
    margin-bottom: 4px;
    width: calc(100% - 64px);

    .sendbird-openchannel-file-message__right__title {
      position: relative;
      display: flex;
      margin-top: 8px;

      .sendbird-openchannel-file-message__right__title__sender-name {
        position: relative;
        display: flex;
      }
      .sendbird-openchannel-file-message__right__title__sent-at {
        position: relative;
        display: flex;
        margin-left: 4px;
      }
    }
    .sendbird-openchannel-file-message__right__body {
      position: relative;
      display: flex;
      margin-top: 4px;
      margin-bottom: 4px;
      border-radius: 8px;
      @include themed() {
        background-color: t(bg-1);
        border: solid 1px t(on-bg-4);
      }

      .sendbird-openchannel-file-message__right__body__icon {
        position: relative;
        display: inline-block;
        margin: 8px;
        border-radius: 8.6px;
        @include themed() {
          background-color: t(bg-0);
        }
      }
      .sendbird-openchannel-file-message__right__body__file-name {
        position: relative;
        display: flex;
        margin-top: 22px;
        margin-right: 16px;
        margin-left: 4px;
        overflow: hidden;
        height: 20px;
        max-height: 20px;
        text-overflow: ellipsis;
        white-space: nowrap;
      }
    }
  }

  .sendbird-openchannel-file-message__context-menu {
    position: absolute;
    display: none;
    top: 8px;
    right: 16px;

    .sendbird-openchannel-file-message__context-menu__icon {
      @include themed() {
        background-color: t(bg--2-5);
      }
      &:hover {
        @include themed() {
          background-color: t(primary--1-bg--7);
        }
      }
    }
  }
}
