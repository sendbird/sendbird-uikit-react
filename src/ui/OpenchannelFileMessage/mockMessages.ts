export const dummyFileMessageImage = {
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1583104552977,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.pdf',
  name: '',
  size: 0,
  type: '',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: 'hoon1234',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

export const dummyFileMessageAudio = {
  messageId: 15814617,
  messageType: 'file',
  channelUrl: 'sendbird_group_channel_11985956_fa67b79d39e6aa5abbeda4413cde24ce1685eedb',
  data: '',
  customType: '',
  createdAt: 1583104552977,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  url: '72b35fef-acda-4e32-123323456789973567890dfvgbhnjmkiytdxbna0e4-37e1226b0615.mp3',
  name: '',
  size: 0,
  type: '',
  thumbnails: [],
  sender: {
    nickname: 'ME',
    profileUrl: 'https://static.sendbird.com/sample/profiles/profile_19_512px.png',
    userId: 'hoon1234',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    isBlockedByMe: false
  },
  reqId: '1583104547956',
  requireAuth: false,
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0
};

type generatorType = (callback?: (message: Record<string, any>) => Record<string, any>) => Record<string, any>;

export const getFileMessage: generatorType = (callback) => {
  const message = { ...dummyFileMessageImage };
  if (callback) {
    return callback(message);
  }
  return message;
};
