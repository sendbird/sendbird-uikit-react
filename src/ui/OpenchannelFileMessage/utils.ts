import { IconTypes } from '../Icon';

export const checkFileType = (fileUrl: string): string => {
  const audioFile = /(\.mp3)$/i;
  const gifFile = /(\.gif)$/i;
  if (audioFile.test(fileUrl)) {
    return IconTypes.FILE_AUDIO;
  }
  if (gifFile.test(fileUrl)) {
    return IconTypes.GIF;
  }
  return IconTypes.FILE_DOCUMENT;
};

export const truncate = (fullStr:string, strLen: number): string => {
  if (fullStr === null || fullStr === undefined) return '';
  if (fullStr.length <= strLen) return fullStr;
  const separator = '...';
  const sepLen = separator.length;
  const charsToShow = strLen - sepLen;
  const frontChars = Math.ceil(charsToShow / 2);
  const backChars = Math.floor(charsToShow / 2);
  return fullStr.substr(0, frontChars) + separator + fullStr.substr(fullStr.length - backChars);
};
