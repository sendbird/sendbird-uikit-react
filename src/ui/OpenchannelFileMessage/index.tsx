import React, { useContext, useRef } from 'react';
import format from 'date-fns/format';
import './index.scss';

import Avatar from '../Avatar';
import ContextMenu, { MenuItems, MenuItem } from '../ContextMenu';
import Label, { LabelTypography, LabelColors } from '../Label';
import Loader from '../Loader';
import Icon, { IconTypes, IconColors } from '../Icon';
import IconButton from '../IconButton';
import TextButton from '../TextButton';
import UserProfile from '../UserProfile';
import { UserProfileContext } from '../../lib/UserProfileContext';

import { LocalizationContext } from '../../lib/LocalizationContext';
import { checkFileType, truncate } from './utils';
import { ClientFileMessage } from '../../index';
import {
  checkIsByMe,
  checkIsPending,
  checkIsFailed,
  isFineDelete,
  isFineResend,
  showMenuTrigger,
} from '../../utils/openChannelUtils';
import { getSenderFromMessage } from '../../utils/openChannelUtils';

interface Props {
  className?: string | Array<string>;
  message: ClientFileMessage;
  userId: string;
  disabled?: boolean;
  chainTop?: boolean;
  chainBottom?: boolean;
  status: string;
  showRemove(bool: boolean): void;
  resendMessage(message: ClientFileMessage): void;
}

export default function OpenchannelFileMessage({
  className,
  message,
  userId,
  disabled,
  chainTop,
  status,
  showRemove,
  resendMessage,
}: Props): JSX.Element {
  const contextMenuRef = useRef(null);
  const avatarRef = useRef(null);
  const { stringSet, dateLocale } = useContext(LocalizationContext);
  const { disableUserProfile, renderUserProfile } = useContext(UserProfileContext);

  const openFileUrl = () => { window.open(message.url); };

  const isByMe = checkIsByMe(message, userId);
  const isPending = checkIsPending(status);
  const isFailed = checkIsFailed(status);
  const sender = getSenderFromMessage(message);

  return (
    <div
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-openchannel-file-message',
      ].join(' ')}
    >
      <div className="sendbird-openchannel-file-message__left">
        {
            !chainTop && (
              <ContextMenu
                menuTrigger={(toggleDropdown) => (
                  <Avatar
                    className="sendbird-openchannel-file-message__left__avatar"
                    src={sender.profileUrl || ''}
                    ref={avatarRef}
                    width="28px"
                    height="28px"
                    onClick={() => {
                      if (!disableUserProfile) {
                        toggleDropdown();
                      }
                    }}
                  />
                )}
                menuItems={(closeDropdown) => (
                  <MenuItems
                    parentRef={avatarRef}
                    parentContainRef={avatarRef}
                    closeDropdown={closeDropdown}
                    style={{ paddingTop: 0, paddingBottom: 0 }}
                  >
                    {
                      renderUserProfile
                      ? (
                        renderUserProfile({
                          user: sender,
                          close: closeDropdown,
                        })
                      )
                      : (
                        <UserProfile
                          user={sender}
                          onSuccess={closeDropdown}
                          disableMessaging
                        />
                      )
                    }
                  </MenuItems>
                )}
              />
            )
          }
      </div>
      <div className="sendbird-openchannel-file-message__right">
        {
          !chainTop && (
            <div className="sendbird-openchannel-file-message__right__title">
              <Label
                className="sendbird-openchannel-file-message__right__title__sender-name"
                type={LabelTypography.CAPTION_2}
                color={isByMe ? LabelColors.SECONDARY_3 : LabelColors.ONBACKGROUND_2}
              >
                {
                  sender && (
                    sender.friendName
                    || sender.nickname
                    || sender.userId
                  )
                }
              </Label>
              <Label
                className="sendbird-openchannel-file-message__right__title__sent-at"
                type={LabelTypography.CAPTION_3}
                color={LabelColors.ONBACKGROUND_3}
              >
                {format(message.createdAt, 'p', { locale: dateLocale })}
              </Label>
            </div>
          )
        }
        <div className="sendbird-openchannel-file-message__right__body">
          {
            checkFileType(message.url) && (
              <Icon
                className="sendbird-openchannel-file-message__right__body__icon"
                type={checkFileType(message.url)}
                fillColor={IconColors.PRIMARY}
                width="48px"
                height="48px"
              />
            )
          }
          <TextButton
            className="sendbird-openchannel-file-message__right__body__file-name"
            onClick={openFileUrl}
          >
            <Label
              type={LabelTypography.BODY_1}
              color={LabelColors.ONBACKGROUND_1}
            >
              {truncate(message.name || message.url, 40)}
            </Label>
          </TextButton>
        </div>
        {
          (isPending || isFailed) && (
            <div className="sendbird-openchannel-file-message__right__tail">
              {
                isPending && (
                  <Loader
                    width="16px"
                    height="16px"
                  >
                    <Icon
                      className="sendbird-openchannel-file-message__right__tail__pending"
                      type={IconTypes.SPINNER}
                      fillColor={IconColors.PRIMARY}
                      width="16px"
                      height="16px"
                    />
                  </Loader>
                )
              }
              {
                isFailed && (
                  <Icon
                    className="sendbird-openchannel-file-message__right__tail__failed"
                    type={IconTypes.ERROR}
                    width="16px"
                    height="16px"
                  />
                )
              }
            </div>
          )
        }
      </div>
      <div
        className="sendbird-openchannel-file-message__context-menu"
        ref={contextMenuRef}
      >
        {
            <ContextMenu
              menuTrigger={(toggleDropdown) => (
                showMenuTrigger({ message, userId, status }) && (
                  <IconButton
                    className="sendbird-openchannel-file-message__context-menu__icon"
                    width="32px"
                    height="32px"
                    onClick={toggleDropdown}
                  >
                    <Icon
                      type={IconTypes.MORE}
                      width="24px"
                      height="24px"
                    />
                  </IconButton>
                )
              )}
              menuItems={(closeDropdown) => (
                <MenuItems
                  parentRef={contextMenuRef}
                  parentContainRef={contextMenuRef}
                  closeDropdown={closeDropdown}
                  openLeft
                >
                  {
                    isFineResend({ message, userId, status }) && (
                      <MenuItem
                        onClick={() => {
                          if (disabled) { return; }
                          resendMessage(message);
                          closeDropdown();
                        }}
                      >
                        {stringSet.CONTEXT_MENU_DROPDOWN__RESEND}
                      </MenuItem>
                    )
                  }
                  {
                    isFineDelete({ message, userId, status }) && (
                      <MenuItem
                        onClick={() => {
                          if (disabled) { return; }
                          showRemove(true);
                          closeDropdown();
                        }}
                      >
                        {stringSet.CONTEXT_MENU_DROPDOWN__DELETE}
                      </MenuItem>
                    )
                  }
                </MenuItems>
              )}
            />
        }
      </div>
    </div>
  );
}
