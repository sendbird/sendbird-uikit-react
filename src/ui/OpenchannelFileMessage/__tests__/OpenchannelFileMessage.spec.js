import React from 'react';
import { render, screen } from '@testing-library/react';
import { dummyFileMessageImage, getFileMessage } from '../mockMessages';

import OpenchannelFileMessage from "../index";

// mock date-fns to avoid problems from snapshot timestamping
// between testing in different locations
// ideally we want to mock date-fns globally - needs more research
jest.mock('date-fns/format', () => () => ('mock-date'));

describe('ui/OpenchannelFileMessage', () => {
  it('should render default elements', function() {
    const { container } = render(
      <OpenchannelFileMessage
        message={dummyFileMessageImage}
      />
    );

    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message')[0].className
    ).toContain('sendbird-openchannel-file-message');
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__left__avatar').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__title__sender-name').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__title__sent-at').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__body__icon').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__body__file-name').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__context-menu').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__tail__pending').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__tail__failed').length
    ).toBe(0);
  });

  it('should not render elements when chainTop is true', function() {
    const { container } = render(
      <OpenchannelFileMessage
        message={dummyFileMessageImage}
        chainTop
      />
    );
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message')[0].className
    ).toContain('sendbird-openchannel-file-message');
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__left__avatar').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__title__sender-name').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__title__sent-at').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__body__icon').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__body__file-name').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__context-menu').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__tail__pending').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__tail__failed').length
    ).toBe(0);
  });

  it('should render pending icon if status is pending', function() {
    const { container } = render(
      <OpenchannelFileMessage
        message={getFileMessage((m) => ({ ...m, sendingStatus: 'pending' }))}
      />
    );
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message')[0].className
    ).toContain('sendbird-openchannel-file-message');
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__left__avatar').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__title__sender-name').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__title__sent-at').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__body__icon').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__body__file-name').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__context-menu').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__tail__pending').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__tail__failed').length
    ).toBe(0);
  });

  it('should render pending icon if status is failed', function() {
    const { container } = render(
      <OpenchannelFileMessage
        message={getFileMessage((m) => ({ ...m, sendingStatus: 'failed' }))}
        status="failed"
      />
    );
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message')[0].className
    ).toContain('sendbird-openchannel-file-message');
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__left__avatar').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__title__sender-name').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__title__sent-at').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__body__icon').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__body__file-name').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__context-menu').length
    ).toBe(1);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__tail__pending').length
    ).toBe(0);
    expect(
      container.getElementsByClassName('sendbird-openchannel-file-message__right__tail__failed').length
    ).toBe(1);
  });

  it('should do a snapshot test of the OpenchannelFileMessage DOM', function() {
    const { asFragment } = render(
      <OpenchannelFileMessage
        message={dummyFileMessageImage}
        status="succeeded"
        userId="hoon1234"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
