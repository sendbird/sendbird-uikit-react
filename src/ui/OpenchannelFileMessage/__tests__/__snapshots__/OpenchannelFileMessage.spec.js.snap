// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/OpenchannelFileMessage should do a snapshot test of the OpenchannelFileMessage DOM 1`] = `
<DocumentFragment>
  <div
    class=" sendbird-openchannel-file-message"
  >
    <div
      class="sendbird-openchannel-file-message__left"
    >
      <div
        class="sendbird-context-menu"
        style="display: inline;"
      >
        <div
          class="sendbird-openchannel-file-message__left__avatar sendbird-avatar"
          role="button"
          style="height: 28px; width: 28px; z-index: 0; left: 0px;"
          tabindex="0"
        >
          <div
            class="sendbird-avatar-img sendbird-image-renderer"
            style="width: 100%; min-width: min(400px, 28px); max-width: 400px;"
          >
            <div
              class="sendbird-image-renderer__image"
              style="width: 100%; min-width: min(400px, 28px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://static.sendbird.com/sample/profiles/profile_19_512px.png);"
            />
            <img
              alt=""
              class="sendbird-image-renderer__hidden-image-loader"
              src="https://static.sendbird.com/sample/profiles/profile_19_512px.png"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="sendbird-openchannel-file-message__right"
    >
      <div
        class="sendbird-openchannel-file-message__right__title"
      >
        <span
          class="sendbird-openchannel-file-message__right__title__sender-name sendbird-label sendbird-label--caption-2 sendbird-label--color-onbackground-2"
        >
          ME
        </span>
        <span
          class="sendbird-openchannel-file-message__right__title__sent-at sendbird-label sendbird-label--caption-3 sendbird-label--color-onbackground-3"
        >
          mock-date
        </span>
      </div>
      <div
        class="sendbird-openchannel-file-message__right__body"
      >
        <div
          class="sendbird-openchannel-file-message__right__body__icon sendbird-icon sendbird-icon-file-document sendbird-icon-color--primary"
          role="button"
          style="width: 48px; min-width: 48px; height: 48px; min-height: 48px;"
          tabindex="0"
        >
          <test-file-stub />
        </div>
        <div
          class="sendbird-openchannel-file-message__right__body__file-name sendbird-color--onbackground-1 sendbird-textbutton "
          role="button"
          tabindex="0"
        >
          <span
            class="sendbird-label sendbird-label--body-1 sendbird-label--color-onbackground-1"
          >
            72b35fef-acda-4e32-...4-37e1226b0615.pdf
          </span>
        </div>
      </div>
    </div>
    <div
      class="sendbird-openchannel-file-message__context-menu"
    >
      <div
        class="sendbird-context-menu"
        style="display: inline;"
      >
        <button
          class="sendbird-openchannel-file-message__context-menu__icon sendbird-iconbutton "
          style="height: 32px; width: 32px;"
          type="button"
        >
          <span
            class="sendbird-iconbutton__inner"
          >
            <div
              class=" sendbird-icon sendbird-icon-more "
              role="button"
              style="width: 24px; min-width: 24px; height: 24px; min-height: 24px;"
              tabindex="0"
            >
              <test-file-stub />
            </div>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
